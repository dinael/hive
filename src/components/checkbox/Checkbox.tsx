import { FC, useState, ChangeEvent } from 'react'
import styles from './Checkbox.module.scss'
import { Label as LabelText } from '../label/Label'
import { Icon } from '../icon/Icon'

export type CheckboxProps = {
  value?: string
  label: string
  id?: string
  required?: boolean
  disabled?: boolean
  className?: string
  onChecked?: (isChecked: boolean) => void
  onError?: (error: boolean) => void
  error: boolean
  errorText?: string
}

export const Checkbox: FC<CheckboxProps> = ({
  label= '',
  id,
  required = false,
  disabled = false,
  className = '',
  onChecked,
  onError,
  error,
  errorText,
  ...props
}) => {

  const [isChecked, setIsChecked] = useState(false)
  const [hasError, setHasError] = useState(false)

  const handleChecked = (e: ChangeEvent<HTMLInputElement>) => {
    setIsChecked(e.target.checked)
    onChecked && onChecked(e.target.checked)
  }

  const handleOnError = () => {
    setHasError(true)
    onError && onError(error)
  }

  return (
      <label
        key={id}
        htmlFor={id}
        className={`
          ${styles['checkbox-component']}
          ${disabled ? styles['disabled'] : ''}
          ${className ? className : ''}
        `}
        {...props}>
        <input
          id={id}
          type="checkbox"
          checked={isChecked}
          onChange={handleChecked}
          onError={handleOnError}
          disabled={disabled}
          aria-checked={isChecked}
          role="checkbox"
          required={required}
          aria-aria-describedby={`validation-message-${id}`}
        />
      <LabelText
        onlyText
        text={label}
      />
      {error ?
        <span
          className={styles['validation-message']}
          role="alert"
          aria-live="assertive"
          aria-atomic="true"
          id={`validation-message-${id}`}>
          <Icon
            name="close"
            size="xs"/>
          {errorText}
        </span>
      : null}
      </label>
  )
}

export default Checkbox
