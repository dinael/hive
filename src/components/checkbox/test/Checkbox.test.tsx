import { describe, test, expect, vi} from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'
import { Checkbox, CheckboxProps } from '../Checkbox'

describe('Checkbox', () => {
  const defaultProps: CheckboxProps = {
    value: 'example-value',
    label: 'Example label',
    error: true,
    errorText: 'Error text',
    id: 'id-test',
    onChecked: vi.fn(),
  }

  test('renders snapshot', () => {
    const { container } = render(<Checkbox {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('renders with the correct label', () => {
    const { container } = render(<Checkbox {...defaultProps} />)
    const labelTextElement = container.querySelector('label > span')
    expect(labelTextElement?.textContent).toBe('Example label')
    expect(container).toMatchSnapshot()
  })

  test('calls onChecked handler when clicked', () => {
    const onCheckedMock = vi.fn()
    const { container } = render(<Checkbox {...defaultProps} onChecked={onCheckedMock} />)
    const checkboxElement = container.querySelector('input')
    if (checkboxElement) {
      fireEvent.change(checkboxElement, { target: { value: 'example-value' } })
    }
    expect(onCheckedMock).toHaveBeenCalledTimes(0)
    expect(container).toMatchSnapshot()
  })

  test('calls onChecked handler when clicked', () => {
    const onCheckedMock = vi.fn()
    const { container } = render(<Checkbox {...defaultProps} onChecked={onCheckedMock} />)
    const checkboxElement = container.querySelector('input')
    if (checkboxElement) {
      fireEvent.change(checkboxElement, { target: { value: 'example-value' } })
    }
    expect(checkboxElement?.value).toBe('example-value')
    expect(container).toMatchSnapshot()
  })
})
