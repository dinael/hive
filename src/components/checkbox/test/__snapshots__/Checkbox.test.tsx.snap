// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > Checkbox calls onChecked handler when clicked 1`] = `
<div>
  <label
    class="
    checkbox-component
    
    
  "
    for="id-test"
    value="example-value"
  >
    <input
      aria-aria-describedby="validation-message-id-test"
      aria-checked="false"
      id="id-test"
      role="checkbox"
      type="checkbox"
      value="example-value"
    />
    <span
      class="
    label-text-component
    
    
    
    
  "
    >
      Example label:
    </span>
    <span
      aria-atomic="true"
      aria-live="assertive"
      class="validation-message"
      id="validation-message-id-test"
      role="alert"
    >
      <span
        aria-hidden="true"
        class="
    icon-component
    xs
    
  "
        style="--icon-color: currentColor;"
      >
        <svg
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            close
          </title>
          <use
            xlink:href="/src/assets/sprite.svg#close"
          />
        </svg>
      </span>
      Error text
    </span>
  </label>
</div>
`;

exports[`Checkbox > Checkbox calls onChecked handler when clicked 2`] = `
<div>
  <label
    class="
    checkbox-component
    
    
  "
    for="id-test"
    value="example-value"
  >
    <input
      aria-aria-describedby="validation-message-id-test"
      aria-checked="false"
      id="id-test"
      role="checkbox"
      type="checkbox"
      value="example-value"
    />
    <span
      class="
    label-text-component
    
    
    
    
  "
    >
      Example label:
    </span>
    <span
      aria-atomic="true"
      aria-live="assertive"
      class="validation-message"
      id="validation-message-id-test"
      role="alert"
    >
      <span
        aria-hidden="true"
        class="
    icon-component
    xs
    
  "
        style="--icon-color: currentColor;"
      >
        <svg
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            close
          </title>
          <use
            xlink:href="/src/assets/sprite.svg#close"
          />
        </svg>
      </span>
      Error text
    </span>
  </label>
</div>
`;

exports[`Checkbox > Checkbox renders snapshot 1`] = `
<div>
  <label
    class="
    checkbox-component
    
    
  "
    for="id-test"
    value="example-value"
  >
    <input
      aria-aria-describedby="validation-message-id-test"
      aria-checked="false"
      id="id-test"
      role="checkbox"
      type="checkbox"
    />
    <span
      class="
    label-text-component
    
    
    
    
  "
    >
      Example label:
    </span>
    <span
      aria-atomic="true"
      aria-live="assertive"
      class="validation-message"
      id="validation-message-id-test"
      role="alert"
    >
      <span
        aria-hidden="true"
        class="
    icon-component
    xs
    
  "
        style="--icon-color: currentColor;"
      >
        <svg
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            close
          </title>
          <use
            xlink:href="/src/assets/sprite.svg#close"
          />
        </svg>
      </span>
      Error text
    </span>
  </label>
</div>
`;

exports[`Checkbox > Checkbox renders with the correct label 1`] = `
<div>
  <label
    class="
    checkbox-component
    
    
  "
    for="id-test"
    value="example-value"
  >
    <input
      aria-aria-describedby="validation-message-id-test"
      aria-checked="false"
      id="id-test"
      role="checkbox"
      type="checkbox"
    />
    <span
      class="
    label-text-component
    
    
    
    
  "
    >
      Example label:
    </span>
    <span
      aria-atomic="true"
      aria-live="assertive"
      class="validation-message"
      id="validation-message-id-test"
      role="alert"
    >
      <span
        aria-hidden="true"
        class="
    icon-component
    xs
    
  "
        style="--icon-color: currentColor;"
      >
        <svg
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <title>
            close
          </title>
          <use
            xlink:href="/src/assets/sprite.svg#close"
          />
        </svg>
      </span>
      Error text
    </span>
  </label>
</div>
`;
